src/GridFS/GridFSAdapter.php

```diff
@@ -0,0 +1,404 @@
+        }
+    }
+
+    public function setVisibility(string $path, string $visibility): void
+    {
+        $file = $this->findFile($path);
+
+        if ($file === null) {
+            throw UnableToSetVisibility::atLocation($path, 'file does not exist');
+        }
+
+        try {
+            $this->bucket->getFilesCollection()->updateOne(
+                ['_id' => $file['_id']],
+                ['$set' => ['metadata.' . self::METADATA_VISIBILITY => $visibility]],
+            );
+        } catch (Exception $exception) {
+            throw UnableToSetVisibility::atLocation($path, $exception->getMessage(), $exception);
+        }
+    }
+
+    public function visibility(string $path): FileAttributes
+    {
+        $file = $this->findFile($path);
+
+        if ($file === null) {
+            throw UnableToRetrieveMetadata::mimeType($path, 'file does not exist');
+        }
+
+        return new FileAttributes($path, null, $file['metadata'][self::METADATA_VISIBILITY] ?? null);
+    }
+
+    public function mimeType(string $path): FileAttributes
+    {
+        if (str_ends_with($path, '/')) {
+            throw UnableToRetrieveMetadata::mimeType($path, 'file path cannot end with a slash');
+        }
+
+        $file = $this->findFile($path);
+
+        if ($file === null) {
+            throw UnableToRetrieveMetadata::mimeType($path, 'file does not exist');
+        }
+
+        if ( ! isset($file['metadata'][self::METADATA_MIMETYPE])) {
+            throw UnableToRetrieveMetadata::mimeType($path, 'unknown');
+        }
+
+        return new FileAttributes($path, null, null, null, $file['metadata'][self::METADATA_MIMETYPE]);
+    }
+
+    public function lastModified(string $path): FileAttributes
+    {
+        if (str_ends_with($path, '/')) {
+            throw UnableToRetrieveMetadata::lastModified($path, 'file path cannot end with a slash');
+        }
+
+        $file = $this->findFile($path);
+
+        if ($file === null) {
+            throw UnableToRetrieveMetadata::lastModified($path, 'file does not exist');
+        }
+
+        return new FileAttributes($path, null, null, $file['uploadDate']->toDateTime()->getTimestamp());
+    }
+
+    public function fileSize(string $path): FileAttributes
+    {
+        if (str_ends_with($path, '/')) {
+            throw UnableToRetrieveMetadata::fileSize($path, 'file path cannot end with a slash');
+        }
+
+        $file = $this->findFile($path);
+
+        if ($file === null) {
+            throw UnableToRetrieveMetadata::fileSize($path, 'file does not exist');
+        }
+
+        return new FileAttributes($path, $file['length']);
+    }
+
+    public function listContents(string $path, bool $deep): iterable
+    {
+        $path = $this->prefixer->prefixDirectoryPath($path);
+
+        $pathdeep = 0;
+        $pipeline = [];
+        if ($path !== '') {
+            $pathdeep = substr_count($path, '/');
+            // Exclude files that does not start with the expected path
```

Comment:
```suggestion
            // Exclude files that do not start with the expected path
```

Subject/verb agreement.